serde/src/ser/mod.rs

```diff
@@ -607,6 +608,39 @@ pub trait Serializer {
         variant: &'static str,
         len: usize,
     ) -> Result<Self::SerializeStructVariant, Self::Error>;
+
+    /// Collect an iterator as a sequence.
+    ///
+    /// The default implementation serializes each item yielded by the iterator
+    /// using `Self::SerializeSeq`. Implementors should not need to override
+    /// this method.
+    fn collect_seq<I>(self, iter: I) -> Result<Self::Ok, Self::Error>
```

Comment:
we can immediately use this in https://github.com/nox/serde/blob/a0a4d782957d351b608d08aa892e7e86317b5182/serde/src/ser/impls.rs#L224

Can we use the same [size-trick](https://github.com/nox/serde/blob/a0a4d782957d351b608d08aa892e7e86317b5182/serde/src/ser/impls.rs#L236) so we don't pass `None`?

Is there a way to use specialization to improve this impl for `ExactSizeIterator`? If so, can this function be written in a way so that is possible (by adding a `feature=unstable` specialization, which gets checked on nightly travis)?